base/compiler/tfuncs.jl

```diff
@@ -406,8 +406,15 @@ function nfields_tfunc(@nospecialize(x))
     x = unwrap_unionall(widenconst(x))
     isconstType(x) && return Const(nfields(x.parameters[1]))
     if isa(x, DataType) && !isabstracttype(x)
-        if !(x.name === Tuple.name && isvatuple(x)) &&
-           !(x.name === _NAMEDTUPLE_NAME && !isconcretetype(x))
+        if x.name === Tuple.name
+            isvatuple(x) && return Int
+            return Const(length(x.types))
+        elseif x.name === _NAMEDTUPLE_NAME
+            length(x.parameters) == 2 || return Int
```

Comment:
rather than falling back to `Int`, fall back to the case we had before in case types are known, but names are not? Probably not a particularly common case, but might as well.

Reply:
Do you mean something like this?
```diff
diff --git a/base/compiler/tfuncs.jl b/base/compiler/tfuncs.jl
index 2675ea4194..6a07f29583 100644
--- a/base/compiler/tfuncs.jl
+++ b/base/compiler/tfuncs.jl
@@ -412,7 +412,7 @@ function nfields_tfunc(@nospecialize(x))
         elseif x.name === _NAMEDTUPLE_NAME
             length(x.parameters) == 2 || return Int
             names = x.parameters[1]
-            isa(names, Tuple{Vararg{Symbol}}) || return Int
+            isa(names, Tuple{Vararg{Symbol}}) || return nfields_tfunc(x.parameters[2])
             return Const(length(names))
         else
             return Const(isdefined(x, :types) ? length(x.types) : length(x.name.names))
diff --git a/test/compiler/inference.jl b/test/compiler/inference.jl
index a87bd2d11b..962a7f2d60 100644
--- a/test/compiler/inference.jl
+++ b/test/compiler/inference.jl
@@ -1529,6 +1529,7 @@ end
 @test nfields_tfunc(@NamedTuple{a::Int,b::Integer}) === Const(2)
 @test nfields_tfunc(NamedTuple{(:a,:b),T} where T<:Tuple{Int,Integer}) === Const(2)
 @test nfields_tfunc(NamedTuple{(:a,:b)}) === Const(2)
+@test nfields_tfunc(NamedTuple{names,Tuple{Any,Any}}) === Const(2)
 @test nfields_tfunc(Union{NamedTuple{(:a,:b)},NamedTuple{(:c,:d)}}) === Const(2)
 
 using Core.Compiler: typeof_tfunc
```

Reply:
sure, that'd work